s3fs/core.py

```diff
@@ -142,7 +169,10 @@ def _ls(self, path, refresh=False):
                     f['Size'] = 0
                     del f['Name']
             else:
-                files = self.s3.list_objects(Bucket=bucket).get('Contents', [])
+                try:
+                    files = self.s3.list_objects(Bucket=bucket).get('Contents', [])
+                except ClientError:
+                    files = []
```

Comment:
You might take all files, and normalize them by split('/', 1), and then take unique entries?

```python
In [1]: fns = ['foo/a', 'foo/bar/a', 'foo/b', 'foo/bar/b']
In [2]: root = 'foo'
In [3]: {'%s/%s' % (root, fn[len(root) + 1:].split('/')[0]) for fn in fns}
Out[3]: {'foo/a', 'foo/b', 'foo/bar'}
```

Perhaps call sorted on the output as well?